### ISSUE TITLE
Time-frequency (joint) scattering example/API support

### ISSUE THREAD

**Issue by OverLordGoldDragon** [kymatio/kymatio#672]
It is my understanding that [Joint Time-Frequency Scattering](https://www.lostanlen.com/wp-content/uploads/2019/12/anden2019tsp.pdf) is 2D scattering with a spin up & down Morlet wavelet filterbank, time & frequency directions decoupled, applied on 1D scattering coefficients. If that's accurate, seems everything's in place to use it given properly configured `filter_bank` (think `L=2` is needed). 

It'd help to have this API-supported with guiding docstrings, or at least an example of how to use `Scattering1D` & `Scattering2D` to do joint scattering.

**Comment by eickenberg** [comment]
I think the 2D wavelets for joint time-frequency scattering look slightly
different from the 2D ones used for images, but yes, in principle this is
the idea.

> It'd help to have this API-supported with guiding docstrings, or at least
an example of how to use Scattering1D & Scattering2D to do joint scattering.

I agree that this would be nice to have.

**Comment by lostanlen** [comment]
Hello @OverLordGoldDragon (i think i know you from librosa btw!)

Thank you for this feature request. You are right in that joint time--frequency scattering (JTFS) is not supported by Kymatio in its current version. The only available implementations are in MATLAB: ScatNet or my implementation (scattering.m) https://github.com/lostanlen/scattering.m

However, as @eickenberg notes, JTFS is not as simple as stacking a layer of Scattering2D on top of a layer of Scattering1D. The 2-D filtering in JTFS should be applied to the un-averaged output of Scattering1D, noted as U1 in the research papers. This unaveraged output is multirate, and thus cannot be used as an input of Scattering2D. Instead, what is necessary is to apply 1-D filtering in time (just like in second-order time scattering) and then immediately apply 1-D filtering in frequency.

This means that it is needed to:
* design a custom filter bank for the frequency domain
* rewrite the depth-first search so that j2 appears in the *outer* loop, not the inner loop
* concatenate paths which share a common j2 along the j1 axis
* take care of the edge case where the frequency filter is a low-pass (phi) whereas the time filter is a band-pass (psi)
* implement spatiotemporal low-pass filtering at the end to ensure invariance to frequency transposition

I have short-term plans to use more JTFS in research so i'd like to captain this effort. It's far from obvious, but it's within reach. However, given the scale of the task, i think it's prematurate to envision merging this to Kymatio as soon as it's ready. I imagine that other admins (@eickenberg et al.) will prefer to see this code develop in a separate repository for the moment. Once we're satisfied with the tool, we may consider including it as part of the "standard library" of Kymatio. (and you all can be sure that i'll be 🥳 -ing that day!)

Thoughts?

**Comment by edouardoyallon** [comment]
I think the first step is to merge filter_bank_1d and filter_bank_2d. Joint is quite simple to implement (both-rototranslation and spiral scattering), it's really a matter of having a clean code, imho.

@OverLordGoldDragon and @lostanlen you're more than welcome to PR this! (I think @eickenberg is/was working on it too)

**Comment by lostanlen** [comment]
> I think the first step is to merge filter_bank_1d and filter_bank_2d.  

if you wish but i don't see how that relates to this issue? There's no actual 2-D filterbank in JTFS. There are three 1-D filterbanks: one for the first layer and two (time and frequency) for the second layer.

roto-translation would be great to have. Spiral is a relatively "niche" improvement on top of JTFS and does not deserve to be included (and i say this all the more comfortably that i'm the one who invented spiral scattering)

**Comment by OverLordGoldDragon** [comment]
I also seek to use this in research and am willing to assist where reasonably able. I'd mainly go off of the MATLAB implementation as I couldn't extract implementation 'fine detail' from the publication (i.e. @lostanlen 's bullet points and more).

@lostanlen We've met when I was rather too new to STFT; indeed the transform's lot more invertible than I imagined. Have a plan in mind with the separate JTFS repo? A Kymatio fork or clone seems best as it contains much reusable code and should ease the subsequent merger.

Re:

 1. **2D wavelets**: right, they're not same as for images, time & frequency dimensions are treated independently, guess this also needs implementing, but I imagine some 1D logic is reusable.
 2. **Feeding Scat1D to 2D**: isn't the multirate discrepancy overcome with a sufficiently large `oversampling`? With max `oversampling` and no averaging it looks like a plain CWT; below I applied it on a linear chirp and compared with [ssqueezepy.cwt](https://github.com/OverLordGoldDragon/ssqueezepy/blob/0.6.0/ssqueezepy/_cwt.py#L10):

<details>
  <summary>Code</summary>

```python
import numpy as np
from ssqueezepy.visuals import imshow
from ssqueezepy import cwt, TestSignals
from kymatio.numpy import Scattering1D

#%%##########################################################################
N = 2**12
tsigs = TestSignals(N=N)
x = tsigs.echirp(N)[0]

#%%##########################################################################
Wx, scales = cwt(x, ('gmw', {'beta': 90}))

#%%###########################################################################
T = x.shape[-1]
J = 11
Q = 16
scattering = Scattering1D(J, T, Q, oversampling=8, average=False,
                          vectorize=False, out_type='list')

#%%###########################################################################
Scx = scattering(x)

#%%###########################################################################
meta = scattering.meta()
order1 = np.where(meta['order'] == 1)
order2 = np.where(meta['order'] == 2)

_Scx1 = np.vstack([Scx[i]['coef'] for i in order1[0] if len(Scx[i]['coef']) == N])

#%%###########################################################################
ikw = dict(abs=1, cmap='jet', ticks=0)
imshow(Wx,    **ikw, title="abs(ssqueezepy.cwt)")
imshow(_Scx1, **ikw, title="abs(kymatio.Scattering1D)")
```

</details>

![image](https://user-images.githubusercontent.com/16495490/109844038-2937dc80-7c65-11eb-86f3-ed346bc4dbef.png)

**Comment by lostanlen** [comment]
>  I imagine some 1D logic is reusable.

yes.


> Feeding Scat1D to 2D: isn't the multirate discrepancy overcome with a sufficiently large oversampling?

You're absolutely right. But without any assumption on the signal, the oversampling should be of the order of J (the number of octaves), which is going to slow down the computation and increase the memory load.

**Comment by lostanlen** [comment]
```
class TimeFrequencyScatteringTorch(ScatteringTorch1D):
    def __init__(
            self,
            J,
            shape,
            Q,
            average=True,
            oversampling=0,
            out_type="array",
            backend="torch"
        ):
        vectorize = True # for compatibility, will be removed in 0.3
        
        # Second-order scattering object for the time variable
        max_order_tm = 2
        ScatteringTorch1D.__init__(
            self, J, shape, Q, max_order_tm, average,
            oversampling, vectorize, out_type, backend)

        # First-order scattering object for the frequency variable
        max_order_fr = 1
        self.sc_freq = ScatteringTorch1D(
            J, shape, Q, max_order_fr, average,
            oversampling, vectorize, out_type, backend)
```

**Comment by edouardoyallon** [comment]
Ahem, maybe you can make a PR on a local fork and we can have a look on it from time to time?

**Comment by OverLordGoldDragon** [comment]
> slow down the computation and increase the memory load.

Right, but I don't suppose there's a way around it, as a single 2D matrix is needed. To this end ssqueezepy's `cwt` code could be reused, vectorized or for-loop, which I'm currently optimizing for both speed and memory (already outdoes Kymatio in both). (Basically since we only do first-order and no downsampling, we do a simple matrix multiply followed by `ifft`).

**Comment by edouardoyallon** [comment]
@OverLordGoldDragon maybe you could try to PR your code in ours, if the speedup is significant? ASV should be supported soon when @janden or @eickenberg will have time to review the corresponding initial PR

**Comment by OverLordGoldDragon** [comment]
@edouardoyallon It's significant but not kymatio-friendly; root of speedup is vectorization, which isn't very compatible with all the downsampling and alternating between first- and second-orders. Other speedup comes from allowing single precision, and `pyfftw` planning (i.e. CPU speedup).  I'm fine for porting it specifically for time-frequency scattering though.

Regarding PR: doable, can keep both a PR and separate repo copy for 'freer' commits. I'm thinking of adding `kymatio/scattering1d_joint` to contain the unique filters, utilities and such - would that work?

**Comment by edouardoyallon** [comment]
> @edouardoyallon It's significant but not kymatio-friendly; root of speedup is vectorization, which isn't very compatible with all the downsampling and alternating between first- and second-orders. Other speedup comes from allowing single precision, and `pyfftw` planning (i.e. CPU speedup). I'm fine for porting it specifically for time-frequency scattering though.
> 
> Regarding PR: doable, can keep both a PR and separate repo copy for 'freer' commits. I'm thinking of adding `kymatio/scattering1d_joint` to contain the unique filters, utilities and such - would that work?

It's better to keep everything separate. We're pretty conservative, for the sake of maintenance, compatibility and code clarity. For instance, for the joint, there is this question like, should we subset in core.py, should we have core_joint.py, ... to me there are 3 steps: a/ filterbanks output should be the same between 1-2D(joint is possible in 2D scattering too) b/ one should create the joint filterbank c/ one should decide how to integrate the joint in a friendly-user manner. In other words, what would be a naive and typicall call from a new user w.r.t. the joint scattering? @eickenberg @janden thoughts?

**Comment by OverLordGoldDragon** [comment]
I'll be needing JTFS soon. What needs doing around the repository for next release? I'm willing to help.

**Comment by lostanlen** [comment]
@OverLordGoldDragon me too; let's team up then!

I have given you push access on lostanlen/kymatio. Feel free to edit my branch fix-674
Lines 118-150 (i.e., convolution along frequency) are still in a haze:
* the IFFT on line 120 should be a forward FFT (with complex input)
* i forgot to take the IFFT after line 127, as you noted
* we need to account for "spin", i.e. upward vs. downard frequency modulation. this can either be done by conjugating the wavelets in the log-frequency-domain filterbank or by conjugating the input and output of convolution. My MATLAB implementation `scattering.m` relies on the former but for Kymatio it would favor the latter because it would require no change in filterbank construction
* we need to account for psi-phi pairs, i.e., when the time-domain filter is a band-pass (psi) and the log-frequency-domain filter (phi) is a low pass.
* we need to agree on a convention for denoting variables in JTFS. Would be good to have @janden's input on this topic at some point.
* there's no unpadding yet

after that we should have a working implementation. a good test is to pass an exponential chirp and see if most of the energy of its JTFS transform is concentrated on either spin. Reversing the time dimension of the input should have the same effect as flipping the spin of JTFS. (because our temporal wavelets have a symmetric amplitude profile)

**Comment by OverLordGoldDragon** [comment]
@lostanlen Sounds good; am wrapping up `ssqueezepy`, will get to it right after. Your bullets are exactly what I figured - is the plan to first implement it as in your branch, then rewrite in `dev`'s flavor? p.s. I'm also willing to work on other stuff, I recall kymatio having bugs in basic `Scattering1D` - just need to know what code is the most concurrent.

**Comment by OverLordGoldDragon** [comment]
Made edits; added entries and made Numpy backend for easier debugging, also bug fixed torch backend. These can be undone in `dev`. Comments/questions:

 1. Backend should provide `zeros` method so we write directly to one array and avoid `concatenate`.
 2. `subsample_fourier` isn't appropriate on tensors shaped `(batch_size, n_freqs, N)`, as it's meant for `(batch_size, 1, N)`. 2D backend's not appropriate either; need new method: `x.reshape(-1, x.shape[1], k, x.shape[-1]//k).mean(axis=-2)`.
 3. Also need frequency-dedicated `ind_start` & `ind_end`
 4. What's the point of `S2_list`?
 5. `.contiguous()` in `transpose` makes a copy, instead `fft` should allow `axis` argument and multiplications be broadcast accordingly
 6. torch backend's `transpose(-2, -3)` will error as number of input args must match input's `ndim`
 7. `padding_row` won't always work, fails here e.g. with `oversampling=100`, also unclear how to manage subsamplings so that `concatenate` remains possible
 8. `get_J_fr()` ends up computing `J` for which `phi_f[k_fr_J]` does not exist, I tried several input `T, J, Q`; this needs to error earlier or need to make said filter

Some edits are temporary, others, along my comments, may be incorrect; can't debug further without resolving 7 & 8. Will continue tomorrow.

<details>
  <summary>debug script</summary>

```python
import numpy as np
from kymatio.numpy import Scattering1D, TimeFrequencyScattering

fn = (Scattering1D, TimeFrequencyScattering)[1]
#%%
T = 1024
J = 6
Q = 12
scattering = fn(J, T, Q)

#%%
x = np.random.randn(T)
Scx = scattering(x)
```
</details>

**Comment by OverLordGoldDragon** [comment]
Strangely didn't realize the coefficients are 2D (thus total 3D), this'll explode output size if not downsampled properly; what's the suggested expedient to inspect these (like in publications)? Here's `imshow(Scx[:, 10])` of a linear chirp from low to Nyquist (made a bunch of temp edits to make it work) - all slices along dim1 look like this, with slight shifts:

<img src="https://user-images.githubusercontent.com/16495490/113033559-3384c200-91a2-11eb-9550-205966683346.png" width=500>

**Comment by lostanlen** [comment]
@OverLordGoldDragon many thanks for your work today.

> 1. Backend should provide zeros method so we write directly to one array and avoid concatenate.

Is concatenation really that much slower? I'm worried about having to maintain a `zeros` method across all backends just for this instance in JTFS. Unless the `zeros` method can serve elsewhere too?


> 2. subsample_fourier isn't appropriate on tensors shaped (batch_size, n_freqs, N), as it's meant for (batch_size, 1, N). 2D backend's not appropriate either; need new method: x.reshape(-1, x.shape[1], k, x.shape[-1]//k).mean(axis=-2).

i imagine the former is a special case of the latter. perhaps we could generalize `subsample_fourier` instead of porting an entirely new method?

> 3. Also need frequency-dedicated ind_start & ind_end

RIght. This is actually one of the subtlest aspects of JTFS, because the ind_end is dependent on `j2`.

> 4. What's the point of S2_list?

Good question. At the moment it's useless. I originally started it for eventually concatenating all JTFS paths and produce a time--frequency--path tensor. But perhaps this can be obtained directly from `out_S2`. I can have it either way. It's more a question of formatting (and memory allocation) than computation itself.

> 5. contiguous() in transpose makes a copy, instead fft should allow axis argument and multiplications be broadcast accordingly

That's a good point. It involves one more modification to the backends, but i'd say it's worth it. @eickenberg @edouardoyallon thoughts on adding a dim kwarg to `fft`?

> 6. 7.

Not sure what these are about but it looks like you're on the right track to get them solved.

> 8. get_J_fr() ends up computing J for which phi_f[k_fr_J] does not exist,

Yes, that was a simplified heuristic which i wrote in a rush. It's important to make it exact.


@OverLordGoldDragon i'm going to be busy with student vivas for the next two days but i'd be happy to sync up on Friday, either here or Discord / Zoom

**Comment by OverLordGoldDragon** [comment]
>  conjugating the wavelets ... or conjugating the input and output of convolution ... Kymatio would favor the latter because it would require no change in filterbank construction

While it's easier to code, two conjugations add non-trivial computational expense. With former the only price is memory, which _might_ be significant; ideally it can be user-settable. I can add latter for starters and bench its expense.

> we need to account for psi-phi pairs, i.e., when the time-domain filter is a band-pass (psi) and the log-frequency-domain filter (phi) is a low pass.

Don't we already? Every psi has a corresponding phi as far as I see.

**Comment by lostanlen** [comment]
> While it's easier to code, two conjugations add non-trivial computational expense. With former the only price is memory, which might be significant; ideally it can be user-settable. I can add latter for starters and bench its expense.

Many thanks. I'll trust your judgment on the computational expense of conjugation. I assumed that it was negligible in front of IFFT ?
The memory cost of conjugating the filters is negligible. There's few of them (<10) and they have a small support (<1000) so their storage footprint is small in comparison with the storage of first-order filters (typically ~100 filters with a support of ~10000)


> Don't we already? Every psi has a corresponding phi as far as I see. -- will address other points later

Yes, the at filterbank construction there is a phi (low-pass filters aka scaling function) associated to every psi (band-pass aka wavelet)
The definition of time scattering is
|| x * psi | * psi | * phi

In JTFS we have a 2-D wavelet at the second order, which we denote by Psi (uppercase to denote that it's 2D) and factorize as a product of two 1-D filters. And a 2-D lowpass filter Phi (again uppercase) factorized as a product of two 1-D low-pass filters.
The definition becomes
|| x * psi | * Psi | * Phi

My point is that the Psi in the equation above may either take the form of a 
(1) Psi = (psi_t  x  psi_f)
or a
(2) Psi = (psi_t  x  phi_f)
or a
(3) Psi = (phi_t x psi_f)

Indeed, if you multiply a band-pass filter on the horizontal dimension (psi_t) with a low-pass filter on the vertical dimension (phi_f), you get a 2-D *band*-pass filter oriented horizontally, i.e. capturing temporal modulations. (much like time scatterig but across several adjacent frequency bands)

Likewise, (phi_t x psi_f) is a 2-D *band*-pass filter oriented vertically.

Lastly, we have Phi = (phi_t x phi_f) as expected.

As the moment, this branch only implements type-1 (psi x psi) coefficients of JTFS. Type-2 (psi x phi) and Type-3 (phi x psi) are missing. Yet they are important as they contain energy and information that isn't present in Type-1 coefficients.

Does that make sense?

**Comment by changhongw** [comment]
@lostanlen @OverLordGoldDragon Thanks so much for launching the implementation of the joint time-frequency scattering! My research focuses on the application of time-frequency scattering to different sound classification problems. All my experiments so far are based on ScatNet. It would be really nice to transfer to Kymatio when the frequential scattering is available. I am currently testing your code and trying to visualise the joint scattering coefficients in a multi-dimensional manner. I will keep you posted about the progress and will try my best to support this implementation!

**Comment by OverLordGoldDragon** [comment]
@changhongw Look forward to your contributions! I've pushed some fixes that should help.

@lostanlen Makes sense, I'll look into it.

**Comment by OverLordGoldDragon** [comment]
Two time-domain conjugations are already 20% of single-threaded FFT's runtime; freq-domain ones (`x[1:] = x[1:][::-1]`) involve rewriting both real and imaginary components and take ~thrice as long, so over half of FFT's time. Basically, FFT is stupid fast.
 
Didn't bench multi-thread or GPU (would have to implement), but I'd not be surprised at similar results - plus we'll have both spin wavelets readily available for visualizing. And you're right, memory's negligible; yeah, I say we pre-build them.

On concatenating, surprisingly the most expensive operation is `0 * array`, can be dozens times slower than `zeros(array.shape)`. Concat isn't much slower because it's ultimately same number of writes, since unfortunately `fft` doesn't provide an `out=`. _However_, `subsample_fourier`, which precedes `fft`, could be implemented to write directly to a pre-assigned `out`. It's also a fairly basic utility, clearer than `0 * arr`, and a one-liner: I advocate for it.

**Comment by lostanlen** [comment]
I see. Many thanks for benchmarking this. So are you saying that we need a `zeros` primitive for initialization?
If so, i can take care of writing a separate PR with:
* the zeros primitive + unit tests
* your transpose primitive + unit tests
and work with the other Kymatio admins towards getting it merged into `dev` quickly. In this way, #674 will become a little bit less heavy in terms of code review.

Meanwhile, perhaps you can work on making the `filter_bank` module allow complex-input filterbanks (i.e. covering the entire frequency axis instead of just the negative frequencies) ?

**Comment by OverLordGoldDragon** [comment]
@lostanlen Am actually close to wrapping up the spins, should push in 30 mins or so

> your transpose primitive

Yet to fully verify if (e.g. for batched inputs), careful there

> making the `filter_bank` module allow complex-input filterbanks 

I don't follow; isn't this what spins do? I implemented spins without touching `filter_bank.py`

**Comment by lostanlen** [comment]
> I don't follow; isn't this what spins do? I implemented spins without touching filter_bank.py

Right, sorry, i was out of phase and hadn't realized that modifications to the `filter_bank` are no longer necessary in light of your spin implementation.

**Comment by rastegah** [comment]
Great to see this is happening and thanks for making this branch available. I have been wondering why Kymatio does not support TF scattering - I am very interested in using this in my research over MATLAB since it is easily integrated with torch

**Comment by OverLordGoldDragon** [comment]
I now advocate for leaving the frequential dimension by default:

 1. The true JTFS is three dimensional and users should be aware of it, and that when it's 2D, a global averaging took place
 2. Most uses should find the frequential axis long enough to contain valuable information worth keeping

Further, I'm unsure for need to implement global averaging, since, the frequential dimension will be the shortest of all by far, so added compute is likely trivial... Can still do but I'd prioritize other things first

> I'm not sure i agree. Presumably, both time-frequency scattering can return a 3-D tensor indexed by time (t), log-frequency (j1) , and scattering modulation (a multiindex combining j2, j_fr, and spin).
> This tensor can readily be plugged into a 2D convnet with spectrotemporal filters. Time and log-frequency would play the role of horizontal and vertical dimensions while scattering modulation would be an unstructured "channel" (like RGB in images).

After understanding JTFS better, I agree: the 2D filter slices are indeed fully suited to be treated independently. And while this'll work, additional considerations:

 3. 3D convs may exploit cross-wavelet dependencies
 4. Multi-channel inputs: it's preferable to reserve input channels for "channels" rather than flattening them

**Comment by lostanlen** [comment]
> 3D convs may exploit cross-wavelet dependencies

yes!! an exciting research perspective indeed (never has a neural network been trained on JTFS coefficients)

> Multi-channel inputs: it's preferable to reserve input channels for "channels" rather than flattening them

You're right. I used the word  "channels" in reference to what a typical `Conv2D` layer would take; but in our case, it's important to separate the JTFS variable dimension (i.e., containing neither time nor frequency but rather the second-order scales and spin) from the audio channel dimension (e..g, left vs right in stereo music).
In full generality, we could have a 5-D output
* time
* frequency
* audio channel (left / right)
* scattering path (scale / rate / spin)
* batch

(i forgot in which order though, so don't quote me on this)

**Comment by Buffy-nevc** [comment]
Hi,

I might be a bit lost over the long trail of discussion. So is JTFS now available in Kymatio? Someone listed a link on this webpage:  https://dsp.stackexchange.com/questions/78622/joint-time-frequency-scattering-explanation

Install as:  pip install git+https://github.com/kymatio/kymatio.git@refs/pull/674/head

Is this what I am suppose to do?

Thanks,

**Comment by rastegah** [comment]
Joint Time-Frequency Scattering is not yet part of the `main` Kymatio branch. We are testing and reviewing it. Soon it will be merged to `dev`, with the intention to release this year.

The current pull request `jtfs-min` supports JTFS via the numpy frontend: `https://github.com/kymatio/kymatio/pull/781`. You can install it with `pip install git+https://github.com/kymatio/kymatio.git@refs/pull/781/head`

If you wish to use the Pytorch or TF frontends, you can use @OverLordGoldDragon's development branch (at your own risk).

`pip install git+git://github.com/lostanlen/kymatio.git@5fa50fb3b718796f3a0e2a97e02cfd07e5d7fd77` 
This is the last commit that passes tests.

**Comment by OverLordGoldDragon** [comment]
@Buffy-nevc Yes, `pip install git+https://github.com/kymatio/kymatio.git@refs/pull/674/head` is fully operational.

@rastegah I ensure my commits to #674 are stable, the dev branches are PRs to it (currently [here](https://github.com/lostanlen/kymatio/pull/27), and also stable, note all 1d pass).

**Comment by Buffy-nevc** [comment]
Hi,

I have tried to install the following in a brand new virtual environment, with no success: 

pip install git+https://github.com/kymatio/kymatio.git@refs/pull/781/head
pip install git+https://github.com/kymatio/kymatio.git@refs/pull/674/head

Error message shown as:
error: pathspec 'refs/pull/674/head did not match any file(s) to git
Command 'git checkout -q refs/pull/674/head' failed with error code 1 in /tmp/pip-a3ekilpo-build 

It seems that git wasn't able to find the file  or path.  Would it be the same if I just download it from:
https://github.com/kymatio/kymatio/pull/674

Thanks,

Buffy

**Comment by OverLordGoldDragon** [comment]
@Buffy-nevc I don't run into this error. [SO](stackoverflow.com/) might help. Alternatively yes, cloning directly works.

**Comment by sammlapp** [comment]
Hi all, from the above I can't tell if frequency-time scattering is now supported in the main (or some feature) branch of this package, and if so, how it can be used. Thanks for any help

**Comment by lostanlen** [comment]
Hello @sammlapp , it's in the main branch, not yet available on PyPI. You have to install from source

**Comment by sammlapp** [comment]
Thanks, I did find it eventually. This is a great tool and API 👏
